src/easylink/step.py

```diff
@@ -1,34 +1,292 @@
+from abc import ABC, abstractmethod
 from dataclasses import dataclass
-from typing import TYPE_CHECKING, Callable
+from pathlib import Path
+from typing import TYPE_CHECKING, Callable, Dict, List, Optional, Tuple
 
-from networkx import MultiDiGraph
+import networkx as nx
+from layered_config_tree import LayeredConfigTree
 
 from easylink.implementation import Implementation
-from easylink.utilities.validation_utils import validate_input_file_dummy
+from easylink.utilities import paths
+from easylink.utilities.data_utils import load_yaml
 
 if TYPE_CHECKING:
     from easylink.configuration import Config
 
 
 @dataclass
-class Step:
+class InputSlot:
+    """StepInput is a dataclass that represents a single input slot for a step."""
+
+    name: str
+    env_var: Optional[str]
+    validator: Callable
+
+
+class Step(ABC):
     """Steps contain information about the purpose of the interoperable elements of
     the sequence called a *Pipeline* and how those elements relate to one another.
     In turn, steps are implemented by Implementations, such that each step may have
     several implementations but each implementation must have exactly one step.
     In a sense, steps contain metadata about the implementations to which they relate.
     """
 
-    name: str
-    prev_input: bool  # Toy version of TODO [MIC-4774]
-    input_files: list  # Toy version of TODO [MIC-4774]
-    input_validator: Callable = validate_input_file_dummy
-
-    def get_subgraph(self, config: "Config") -> MultiDiGraph:
-        sub = MultiDiGraph()
-        sub.add_node(
-            config.schema.get_step_id(self),
-            implementation=Implementation(config=config, step=self),
-            input_validator=self.input_validator,
+    def __init__(
+        self, name: str, input_slots: List[Tuple[str]] = [], output_slots: List[str] = []
+    ) -> None:
+        self.name = name
+        self.input_slots = {slot[0]: InputSlot(*slot) for slot in input_slots}
+        self.output_slots = output_slots
+
+    @abstractmethod
+    def get_implementation_graph(self, graph, pipeline_config) -> nx.MultiDiGraph:
```

Comment:
Steps must be able to:

- produce a graph of nodes composed of implementations
- update the edges of the graph from the step graph to the implementation graph
- perform validation against the config